src/ray/core_worker/object_interface.cc

```diff
@@ -80,24 +94,107 @@ Status CoreWorkerObjectInterface::Get(
         ids, timeout_ms, worker_context_.GetCurrentTaskID(), &objects));
     RAY_CHECK(ids.size() == objects.size());
     for (size_t i = 0; i < objects.size(); i++) {
-      (*results).emplace(ids[i], objects[i]);
+      if (objects[i] != nullptr) {
+        (*results).emplace(ids[i], objects[i]);
+      }
+    }
+  }
+
+  return Status::OK();
+}
+
+Status CoreWorkerObjectInterface::Wait(const std::vector<ObjectID> &ids, int num_objects,
+                                       int64_t timeout_ms, std::vector<bool> *results) {
+  (*results).resize(ids.size(), false);
+
+  if (num_objects <= 0 || num_objects > ids.size()) {
+    return Status::Invalid("num_objects value is not valid");
+  }
+
+  // There can be a few cases here:
+  // - for task return objects, find the store provider type for an object from
+  //   its transport, and then try to get from the corresponding store provider;
+  // - for other objects, try to get from plasma store.
+  EnumUnorderedMap<StoreProviderType, std::unordered_set<ObjectID>>
+      object_ids_per_store_provider;
+  GetObjectIdsPerStoreProvider(ids, &object_ids_per_store_provider);
+
+  std::unordered_map<ObjectID, int> object_counts;
+  for (const auto &entry : ids) {
+    auto iter = object_counts.find(entry);
+    if (iter == object_counts.end()) {
+      object_counts.emplace(entry, 1);
+    } else {
+      iter->second++;
+    }
+  }
+
+  auto current_timeout_ms = timeout_ms;
+
+  for (const auto &entry : object_ids_per_store_provider) {
```

Comment:
Hmmm, in this case we will use up the entire timeout on trying to get objects from one store provider before even trying another (which might have all of the objects available). Maybe we should try all of them with timeout == 0 first and then move to the timeout logic?

Reply:
good point. thanks